flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/TwoStageOptimizedWindowAggregateRule.java

```diff
@@ -0,0 +1,168 @@
+ */
+public class TwoStageOptimizedWindowAggregateRule extends RelOptRule {
+
+    public static final TwoStageOptimizedWindowAggregateRule INSTANCE =
+            new TwoStageOptimizedWindowAggregateRule();
+
+    private TwoStageOptimizedWindowAggregateRule() {
+        super(
+                operand(
+                        StreamPhysicalWindowAggregate.class,
+                        operand(StreamPhysicalExchange.class, operand(RelNode.class, any()))),
+                "TwoStageOptimizedWindowAggregateRule");
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        final StreamPhysicalWindowAggregate windowAgg = call.rel(0);
+        final RelNode realInput = call.rel(2);
+        final TableConfig tableConfig = unwrapContext(call.getPlanner()).getTableConfig();
+        final WindowingStrategy windowing = windowAgg.windowing();
+
+        // the two-phase optimization must be enabled
+        if (getAggPhaseStrategy(tableConfig) == AggregatePhaseStrategy.ONE_PHASE) {
+            return false;
+        }
+
+        // processing time window doesn't support two-phase,
+        // otherwise the processing-time can't be materialized in a single node
+        if (!windowing.isRowtime()) {
+            return false;
+        }
+
+        // all aggregate function should support merge() method
+        if (!AggregateUtil.doAllSupportPartialMerge(windowAgg.aggInfoList().aggInfos())) {
+            return false;
+        }
+
+        return !isInputSatisfyRequiredDistribution(realInput, windowAgg.grouping());
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final StreamPhysicalWindowAggregate windowAgg = call.rel(0);
+        final RelNode realInput = call.rel(2);
+
+        RelTraitSet localTraitSet =
+                realInput
+                        .getTraitSet()
+                        .plus(ModifyKindSetTrait.INSERT_ONLY())
+                        .plus(UpdateKindTrait.NONE());
+        StreamPhysicalLocalWindowAggregate localAgg =
+                new StreamPhysicalLocalWindowAggregate(
+                        windowAgg.getCluster(),
+                        localTraitSet,
+                        realInput,
+                        windowAgg.grouping(),
+                        windowAgg.aggCalls(),
+                        windowAgg.windowing());
+
+        // grouping keys is forwarded by local agg, use indices instead of groupings
+        int[] globalGrouping = IntStream.range(0, windowAgg.grouping().length).toArray();
+        FlinkRelDistribution globalDistribution = createDistribution(globalGrouping);
+        // create exchange if needed
+        RelNode newInput =
+                FlinkExpandConversionRule.satisfyDistribution(
+                        FlinkConventions.STREAM_PHYSICAL(), localAgg, globalDistribution);
+        RelTraitSet globalAggProvidedTraitSet = windowAgg.getTraitSet();
+        SliceAttachedWindowingStrategy globalWindowing =
```

Comment:
It's better to distinguish two different cases here:
1. If `windowStrategy` of original `StreamPhysicalWindowAggregate` is `TimeAttributeWindowingStrategy`,
then `WindowingStrategy` of equivalent `StreamPhysicalGlobalWindowAggregate` is `SliceAttachedWindowingStrategy` like the current code.
2. If `WindowingStrategy` of original `StreamPhysicalWindowAggregate` is `WindowAttachedWindowingStrategy`, then `WindowingStrategy` of equivalent `StreamPhysicalGlobalWindowAggregate` could still be `WindowAttachedWindowingStrategy`. Besides, there is no need to add `sliceEnd` in `StreamPhysicalLocalWindowAggregate`.